packages/flutter/lib/src/widgets/_web_image_web.dart

```diff
@@ -0,0 +1,384 @@
+  static bool _registered = false;
+
+  static void _register() {
+    assert(!_registered);
+    _registered = true;
+    ui_web.platformViewRegistry.registerViewFactory(_viewType, (int viewId,
+        {Object? params}) {
+      final Map<Object?, Object?> paramsMap = params! as Map<Object?, Object?>;
+      // Create a new <img> element. The browser is able to display the image
+      // without fetching it over the network again.
+      final web.HTMLImageElement img =
+          web.document.createElement('img') as web.HTMLImageElement;
+      img.src = paramsMap['src']! as String;
+      return img;
+    });
+  }
+
+  /// The `src` url for the `<img>` tag.
+  final String? src;
+
+  @override
+  Widget build(BuildContext context) {
+    if (src == null) {
+      return const SizedBox.expand();
+    }
+    return HtmlElementView(
+      viewType: _viewType,
+      creationParams: <String, String?>{'src': src},
+    );
+  }
+}
+
+/// A widget which displays and lays out an underlying `<img>` platform view.
+class RawWebImage extends SingleChildRenderObjectWidget {
+  /// Creates a [RawWebImage].
+  RawWebImage({
+    super.key,
+    required this.image,
+    this.debugImageLabel,
+    this.width,
+    this.height,
+    this.fit,
+    this.alignment = Alignment.center,
+    this.matchTextDirection = false,
+  }) : super(child: ImgElementPlatformView(image.htmlImage.src));
+
+  /// The underlying `<img>` element to be displayed.
+  final WebImageInfo image;
+
+  /// A debug label explaining the image.
+  final String? debugImageLabel;
+
+  /// The requested width for this widget.
+  final double? width;
+
+  /// The requested height for this widget.
+  final double? height;
+
+  /// How the `<img>` should be inscribed in the box constraining it.
+  final BoxFit? fit;
+
+  /// How the image should be aligned in the box constraining it.
+  final AlignmentGeometry alignment;
+
+  /// Whether or not the alignment of the image should match the text direction.
+  final bool matchTextDirection;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return RenderWebImage(
+      image: image.htmlImage,
+      width: width,
+      height: height,
+      fit: fit,
+      alignment: alignment,
+      matchTextDirection: matchTextDirection,
+      textDirection: matchTextDirection || alignment is! Alignment
+          ? Directionality.of(context)
+          : null,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, RenderWebImage renderObject) {
+    renderObject
+      ..image = image.htmlImage
+      ..width = width
+      ..height = height
+      ..fit = fit
+      ..alignment = alignment
+      ..matchTextDirection = matchTextDirection
+      ..textDirection = matchTextDirection || alignment is! Alignment
+          ? Directionality.of(context)
+          : null;
+  }
+}
+
+/// Lays out and positions the child `<img>` element similarly to [RenderImage].
+class RenderWebImage extends RenderShiftedBox {
```

Comment:
Did `FittedBox` turn out insufficient? I can't tell if this custom render object provides extra secret sauce compared to using an off-the-shelf layout widget.

Reply:
There are some things that this render object needs to do that the `FittedBox` can't, viz. taking textDirection into account and overriding the "preserve aspect ratio" layout constraints if either width or height are specified.